File: ios/Runner/BallYoloDetector.swift
import Foundation
import TensorFlowLite

// ============================================================
// YOLOv8n int8 TFLite 高爾夫球偵測器（iOS Swift 側）
//
// 模型規格（golfballyolov8n_int8.tflite）：
//   Input : [1, 640, 640, 3] INT8  (zero_point=-128, scale=1/255)
//   Output: [1, 5, 8400]    FLOAT32 or INT8
//     channels: cx, cy, w, h (in 640×640 pixels), conf (0–1)
//
// 對應 Android 側的 BallYoloDetector.kt。
// ============================================================

final class BallYoloDetector {

    // ── 單例 ──────────────────────────────────────────────────
    static let shared = BallYoloDetector()
    private init() {}

    // ── 常數 ──────────────────────────────────────────────────
    static let inputSize            = 640
    static let confThreshold: Float  = 0.25          // 預設門櫛（對應 Android CONF_THRESHOLD）
    private static let nmsIoU: Float = 0.45
    private static let tileEdgeMargin: Float = 24    // tile 邊緣排除（對應 TILE_EDGE_MARGIN）

    // ── 狀態 ──────────────────────────────────────────────────
    private let lock            = NSLock()
    private var interpreter:    Interpreter?
    private var loadAttempted   = false
    // 輸入 tensor 型態
    private var inIsFloat       = false
    private var inZeroPoint:    Int   = -128
    // 輸出 tensor 量化參數
    private var outIsFloat      = true
    private var outScale:       Float = 1.0
    private var outZeroPoint:   Int   = 0

    var isLoaded: Bool { interpreter != nil }

    // ── 偵測結果型別 ──────────────────────────────────────────
    struct Detection {
        let cx, cy, bboxW, bboxH: Int
        let conf: Float
    }

    // MARK: - 載入模型

    /// 嘗試從 Flutter assets 載入模型，只執行一次。
    /// - Returns: 是否成功載入
    @discardableResult
    func tryLoad() -> Bool {
        lock.lock(); defer { lock.unlock() }
        guard !loadAttempted else { return isLoaded }
        loadAttempted = true

        // Flutter assets 在 iOS bundle 內路徑為 flutter_assets/...
        guard let modelPath = Bundle.main.path(
            forResource: "golfballyolov8n_int8",
            ofType: "tflite",
            inDirectory: "flutter_assets/models"
        ) else {
            print("[BallYoloDetector] ❌ 模型 asset 未找到")
            return false
        }

        do {
            var opts = Interpreter.Options()
            opts.threadCount = 2
            let interp = try Interpreter(modelPath: modelPath, options: opts)
            try interp.allocateTensors()

            // 讀取輸入 tensor 型態（FLOAT32 or INT8）
            let inTensor = try interp.input(at: 0)
            inIsFloat = (inTensor.dataType == .float32)
            if !inIsFloat, let qp = inTensor.quantizationParameters {
                inZeroPoint = qp.zeroPoint
            }
            let inType = inIsFloat ? "FLOAT32" : "INT8"
            print("[BallYoloDetector] input: \(inType) zp=\(inZeroPoint)")

            // 讀取輸出量化參數
            let outTensor = try interp.output(at: 0)
            outIsFloat = (outTensor.dataType == .float32)
            if !outIsFloat, let qp = outTensor.quantizationParameters {
                outScale     = qp.scale
                outZeroPoint = qp.zeroPoint
            }
            interpreter = interp
            let shape = outTensor.shape.dimensions
            print("[BallYoloDetector] ✅ 模型載入成功  output=\(shape)  outFloat=\(outIsFloat)")
            return true
        } catch {
            print("[BallYoloDetector] ❌ 模型載入失敗: \(error)")
            return false
        }
    }

    // MARK: - 推論

    /// 對一幀 BGRA 像素執行 YOLOv8 偵測
    ///
    /// - Parameters:
    ///   - bgraBase:  CVPixelBuffer BGRA 基址（已 lock）
    ///   - bytesPerRow: CVPixelBuffer 每列位元組數（含 stride padding）
    ///   - frameW:    display-space 影像寬度
    ///   - frameH:    display-space 影像高度
    /// - Returns:     偵測結果（座標在 display-space）
    /// 對 [roiCenterX, roiCenterY] 附近裁切 640×640 patch 並執行推論。
    /// 對應 Android BallYoloDetector.detect()。
    func detect(
        bgraBase:      UnsafePointer<UInt8>,
        bytesPerRow:   Int,
        frameW:        Int,
        frameH:        Int,
        roiCenterX:    Int   = -1,
        roiCenterY:    Int   = -1,
        confThreshold: Float = BallYoloDetector.confThreshold
    ) -> [Detection] {
        guard let interp = interpreter else { return [] }
        let S = BallYoloDetector.inputSize

        // ── 計算 tile 左上角（640×640 crop，以 ROI 中心為中心）────
        let halfSize = S / 2
        let roiX = roiCenterX >= 0 ? roiCenterX : frameW / 2
        let roiY = roiCenterY >= 0 ? roiCenterY : frameH / 2

        let tileLeft: Int
        let tileTop:  Int
        let tileW:    Int
        let tileH:    Int

        if frameW < S || frameH < S {
            tileLeft = 0; tileTop = 0
            tileW = frameW; tileH = frameH
        } else {
            tileLeft = max(0, min(roiX - halfSize, frameW - S))
            tileTop  = max(0, min(roiY - halfSize, frameH - S))
            tileW = S; tileH = S
        }

        let scaleX = Float(tileW) / Float(S)
        let scaleY = Float(tileH) / Float(S)

        // ── 填入 [1, 640, 640, 3] input ──────────────────────────
        do {
            if inIsFloat {
                // FLOAT32：BGRA → RGB pixel/255
                var inputData = Data(count: S * S * 3 * 4)
                inputData.withUnsafeMutableBytes { (ptr: UnsafeMutableRawBufferPointer) in
                    guard let base = ptr.baseAddress?.assumingMemoryBound(to: Float.self) else { return }
                    for y in 0 ..< S {
                        let srcY = min(tileTop + Int(Float(y) * scaleY), frameH - 1)
                        let rowOff = srcY * bytesPerRow
                        for x in 0 ..< S {
                            let srcX = min(tileLeft + Int(Float(x) * scaleX), frameW - 1)
                            let off = rowOff + srcX * 4
                            let b = Float(bgraBase[off]) / 255.0
                            let g = Float(bgraBase[off + 1]) / 255.0
                            let r = Float(bgraBase[off + 2]) / 255.0
                            let idx = (y * S + x) * 3
                            base[idx] = r
                            base[idx + 1] = g
                            base[idx + 2] = b
                        }
                    }
                }
                try interp.copy(inputData, toInputAt: 0)
            } else {
                // INT8：亮度量化 (luma + zeroPoint) → [-128, 127]
                var inputBytes = [Int8](repeating: Int8(clamping: inZeroPoint), count: S * S * 3)
                for y in 0 ..< S {
                    let srcY = min(tileTop + Int(Float(y) * scaleY), frameH - 1)
                    let rowOff = srcY * bytesPerRow
                    for x in 0 ..< S {
                        let srcX = min(tileLeft + Int(Float(x) * scaleX), frameW - 1)
                        let off = rowOff + srcX * 4
                        let b = Int(bgraBase[off])
                        let g = Int(bgraBase[off + 1])
                        let r = Int(bgraBase[off + 2])
                        let luma = (299 * r + 587 * g + 114 * b) / 1000
                        let q = Int8(clamping: luma + inZeroPoint)
                        let idx = (y * S + x) * 3
                        inputBytes[idx] = q
                        inputBytes[idx + 1] = q
                        inputBytes[idx + 2] = q
                    }
                }
                try interp.copy(Data(bytes: inputBytes, count: inputBytes.count), toInputAt: 0)
            }

            try interp.invoke()
            return parseOutput(
                try interp.output(at: 0),
                tileLeft: tileLeft,
                tileTop: tileTop,
                scaleX: scaleX,
                scaleY: scaleY,
                confThreshold: confThreshold
            )
        } catch {
            print("[BallYoloDetector] 推理失敗: \(error)")
            return []
        }
    }

    // MARK: - 解析輸出 [1, 5, 8400]

    private func parseOutput(
        _ tensor:      Tensor,
        tileLeft:      Int,
        tileTop:       Int,
        scaleX:        Float,
        scaleY:        Float,
        confThreshold: Float
    ) -> [Detection] {
        let S      = BallYoloDetector.inputSize
        let MARGIN = BallYoloDetector.tileEdgeMargin
        let n      = 8400
        let bytes  = tensor.data

        func getVal(ch: Int, anchor: Int) -> Float {
            if outIsFloat {
                return bytes.withUnsafeBytes { (rawBuffer: UnsafeRawBufferPointer) in
                    rawBuffer.load(fromByteOffset: (ch * n + anchor) * 4, as: Float.self)
                }
            } else {
                let raw: UInt8 = bytes.withUnsafeBytes { (rawBuffer: UnsafeRawBufferPointer) in
                    rawBuffer.load(fromByteOffset: ch * n + anchor, as: UInt8.self)
                }
                return Float(Int(raw) - outZeroPoint) * outScale
            }
        }

        // 偵測座標格式（normalized [0,1] 或 pixel [0,640]），對應 Android coordScale 邏輯
        var maxConf: Float = 0
        var maxCx:   Float = 0
        for i in 0 ..< n {
            let c = getVal(ch: 4, anchor: i)
            if c > maxConf { maxConf = c; maxCx = getVal(ch: 0, anchor: i) }
        }
        let coordScale: Float = (maxConf > 0.05 && maxCx < 2.0) ? Float(S) : 1.0

        // 篩選 + tile 邊緣排除 + 座標轉回 frame 空間
        var raw: [(cx: Float, cy: Float, w: Float, h: Float, conf: Float)] = []
        for i in 0 ..< n {
            let conf = getVal(ch: 4, anchor: i)
            guard conf >= confThreshold else { continue }

            let cx_tile = getVal(ch: 0, anchor: i) * coordScale
            let cy_tile = getVal(ch: 1, anchor: i) * coordScale
            let w_tile  = getVal(ch: 2, anchor: i) * coordScale
            let h_tile  = getVal(ch: 3, anchor: i) * coordScale

            if cx_tile < MARGIN || cy_tile < MARGIN ||
               cx_tile > Float(S) - MARGIN || cy_tile > Float(S) - MARGIN {
                continue
            }

            raw.append((
                cx:   Float(tileLeft) + cx_tile * scaleX,
                cy:   Float(tileTop)  + cy_tile * scaleY,
                w:    w_tile * scaleX,
                h:    h_tile * scaleY,
                conf: conf
            ))
        }

        return nms(raw).map { d in
            Detection(
                cx:    Int(d.cx),
                cy:    Int(d.cy),
                bboxW: max(1, Int(d.w)),
                bboxH: max(1, Int(d.h)),
                conf:  d.conf
            )
        }
    }

    // MARK: - NMS

    private func nms(
        _ dets: [(cx: Float, cy: Float, w: Float, h: Float, conf: Float)]
    ) -> [(cx: Float, cy: Float, w: Float, h: Float, conf: Float)] {
        let sorted   = dets.sorted { $0.conf > $1.conf }
        var suppress = [Bool](repeating: false, count: sorted.count)
        var kept     = [(cx: Float, cy: Float, w: Float, h: Float, conf: Float)]()
        for i in sorted.indices {
            guard !suppress[i] else { continue }
            kept.append(sorted[i])
            for j in (i + 1) ..< sorted.count {
                if !suppress[j] && iou(sorted[i], sorted[j]) > BallYoloDetector.nmsIoU {
                    suppress[j] = true
                }
            }
        }
        return kept
    }

    private func iou(
        _ a: (cx: Float, cy: Float, w: Float, h: Float, conf: Float),
        _ b: (cx: Float, cy: Float, w: Float, h: Float, conf: Float)
    ) -> Float {
        let ax1 = a.cx - a.w / 2, ay1 = a.cy - a.h / 2
        let ax2 = a.cx + a.w / 2, ay2 = a.cy + a.h / 2
        let bx1 = b.cx - b.w / 2, by1 = b.cy - b.h / 2
        let bx2 = b.cx + b.w / 2, by2 = b.cy + b.h / 2
        let ix1 = max(ax1, bx1), iy1 = max(ay1, by1)
        let ix2 = min(ax2, bx2), iy2 = min(ay2, by2)
        guard ix2 > ix1 && iy2 > iy1 else { return 0 }
        let inter = (ix2 - ix1) * (iy2 - iy1)
        let aArea = (ax2 - ax1) * (ay2 - ay1)
        let bArea = (bx2 - bx1) * (by2 - by1)
        return inter / (aArea + bArea - inter)
    }
}

File: android/app/src/main/kotlin/com/aethertek/orvia/BallYoloDetector.kt
package com.aethertek.orvia

import android.content.res.AssetManager
import android.util.Log
import com.aethertek.orvia.BuildConfig
import org.tensorflow.lite.DataType
import org.tensorflow.lite.Interpreter
import java.nio.ByteBuffer
import java.nio.ByteOrder
import kotlin.math.max
import kotlin.math.min

/**
 * YOLOv8n TFLite 高爾夫球偵測器。
 *
 * 模型規格（golfballyolov8n_int8.tflite — 實際 FLOAT32 export）：
 *   Input : [1, 640, 640, 3] FLOAT32，pixel/255（0~1 normalized）
 *   Output: [1, 5, 8400] FLOAT32
 *     channels: cx(px), cy(px), w(px), h(px), conf — 相對於 640×640 input
 *
 * 此模型以 SAHI 訓練，每次推論只接受 640×640 的「裁切區域」。
 * 呼叫端（BallYoloExtractor）負責維護 ROI 中心，並傳入 roiCenterX / roiCenterY。
 *
 * detect() 回傳座標已轉換回原始 frame 座標空間。
 */
class BallYoloDetector(private val assetManager: AssetManager) {

    companion object {
        private const val TAG = "BallYoloDetector"
        private const val MODEL_ASSET =
            "flutter_assets/assets/models/golfballyolov8n_int8.tflite"
        const val INPUT_SIZE     = 640
        private const val CONF_THRESHOLD = 0.25f
        private const val NMS_IOU        = 0.45f
        // tile 邊緣 margin：排除距邊界太近的偵測（false positive 常出現在 tile 角落）
        private const val TILE_EDGE_MARGIN = 24f  // pixels in tile space (640×640)
    }

    private var interpreter: Interpreter? = null

    // 輸入 tensor 型態
    private var inIsFloat   = false
    private var inZeroPoint = -128

    // 輸出 tensor 量化參數
    private var outIsFloat   = true
    private var outScale     = 1.0f
    private var outZeroPoint = 0
    private var detectionOutputIdx = 0

    // 可重用 buffer（tryLoad 後依實際型態分配）
    private var inputBuf:  ByteBuffer? = null
    private var outputBuf: ByteBuffer? = null

    // debug 計數
    private var detectCallCount = 0

    // coordScale 快取：模型格式不會改變，第一次確認後不再掃 8400 anchors
    private var cachedCoordScale: Float? = null

    val isLoaded: Boolean get() = interpreter != null

    // ──────────────────────────────────────────────────────────────

    fun tryLoad(): Boolean {
        return try {
            val bytes = assetManager.open(MODEL_ASSET).use { it.readBytes() }
            val modelBuf = ByteBuffer.allocateDirect(bytes.size).apply {
                order(ByteOrder.nativeOrder()); put(bytes); rewind()
            }
            val opts = Interpreter.Options().apply { setNumThreads(4) }
            interpreter = Interpreter(modelBuf, opts)

            // 讀取輸入型態 → 決定 buffer 大小與填值方式
            interpreter!!.getInputTensor(0).also { t ->
                inIsFloat   = (t.dataType() == DataType.FLOAT32)
                inZeroPoint = t.quantizationParams().zeroPoint
                val elemBytes = if (inIsFloat) 4 else 1
                Log.d(TAG, "Input shape=${t.shape().toList()} type=${t.dataType()} " +
                    "inIsFloat=$inIsFloat elemBytes=$elemBytes")
                inputBuf = ByteBuffer.allocateDirect(INPUT_SIZE * INPUT_SIZE * 3 * elemBytes)
                    .apply { order(ByteOrder.nativeOrder()) }
            }

            // 列出所有 output tensors（debug）
            val outCount = interpreter!!.outputTensorCount
            Log.d(TAG, "=== YOLOv8 outputs: count=$outCount ===")
            for (i in 0 until outCount) {
                val t = interpreter!!.getOutputTensor(i)
                Log.d(TAG, "  out[$i] shape=${t.shape().toList()} type=${t.dataType()} " +
                    "scale=${t.quantizationParams().scale} zp=${t.quantizationParams().zeroPoint}")
            }

            // 找到 [*, 5, 8400] 偵測 tensor
            findDetectionOutputTensor()?.also { (idx, tensor) ->
                detectionOutputIdx = idx
                outIsFloat   = tensor.dataType() == DataType.FLOAT32
                outScale     = tensor.quantizationParams().scale
                outZeroPoint = tensor.quantizationParams().zeroPoint
                val elemBytes = if (outIsFloat) 4 else 1
                outputBuf = ByteBuffer.allocateDirect(5 * 8400 * elemBytes)
                    .apply { order(ByteOrder.nativeOrder()) }
                Log.d(TAG, "Detection output[$idx] type=${tensor.dataType()} outIsFloat=$outIsFloat")
            } ?: run {
                Log.w(TAG, "Cannot find detection output tensor")
                interpreter?.close(); interpreter = null; return false
            }

            Log.d(TAG, "YOLOv8 loaded OK " +
                "(in=${if (inIsFloat) "FLOAT32" else "INT8"} " +
                "out=${if (outIsFloat) "FLOAT32" else "INT8"})")
            true
        } catch (e: Exception) {
            Log.w(TAG, "Cannot load YOLOv8 model: $e"); false
        }
    }

    // ──────────────────────────────────────────────────────────────

    /**
     * 對 [roiCenterX, roiCenterY] 附近裁切出 640×640 patch 並執行推論。
     *
     * @param roiCenterX      ROI 中心 X（frame 座標）
     * @param roiCenterY      ROI 中心 Y（frame 座標）
     * @param tileEdgeMargin  tile 邊緣排除距離（pixels in tile space）；
     *                        預設 24f；擊球後可降低至 8f 讓高速球靠近邊緣也能通過
     * @return 偵測列表，FloatArray(5) = [cx_frame, cy_frame, w_px, h_px, conf]
     *         座標已轉換回原始 frame 座標空間
     */
    fun detect(
        yData:   ByteArray, yStride: Int,
        uData:   ByteArray, uStride: Int, uPixelStride: Int,
        vData:   ByteArray, vStride: Int, vPixelStride: Int,
        frameW: Int,
        frameH: Int,
        roiCenterX: Int,
        roiCenterY: Int,
        // 允許呼叫端動態調整：擊球後可降低至 0.05f 讓微小/高速球也能通過
        confThreshold: Float = CONF_THRESHOLD,
        // 擊球後可降低至 8f，避免高速球在 tile 邊緣被過濾
        tileEdgeMargin: Float = TILE_EDGE_MARGIN,
    ): List<FloatArray> {
        val interp = interpreter ?: return emptyList()
        val inBuf  = inputBuf   ?: return emptyList()
        val outBuf = outputBuf  ?: return emptyList()

        // ── 計算 tile 左上角（640×640 crop，以 ROI 中心為中心）────────
        val halfSize = INPUT_SIZE / 2   // 320
        val tileLeft: Int
        val tileTop: Int
        val tileW: Int
        val tileH: Int

        if (frameW < INPUT_SIZE || frameH < INPUT_SIZE) {
            // frame 比模型輸入小 → 全 frame 縮放（保留相容性）
            tileLeft = 0; tileTop = 0
            tileW = frameW; tileH = frameH
        } else {
            // 正常情況：裁 640×640，對齊 frame 邊界
            tileLeft = (roiCenterX - halfSize).coerceIn(0, frameW - INPUT_SIZE)
            tileTop  = (roiCenterY - halfSize).coerceIn(0, frameH - INPUT_SIZE)
            tileW = INPUT_SIZE; tileH = INPUT_SIZE
        }

        // ── 填入 [1, 640, 640, 3] input（Native C 加速）─────────────
        // 原 JVM 迴圈 640×640=40萬次，由 NativeLib.fillYoloInput 替代。
        // 使用 GetDirectBufferAddress 直接寫入 inBuf（ByteBuffer.allocateDirect），零複製。
        inBuf.rewind()
        NativeLib.fillYoloInput(
            yData, yStride,
            uData, uStride, uPixelStride,
            vData, vStride, vPixelStride,
            frameW, frameH,
            tileLeft, tileTop, tileW, tileH,
            INPUT_SIZE,
            inBuf,
            inIsFloat,
            inZeroPoint,
        )
        inBuf.rewind()

        // ── 推論 ─────────────────────────────────────────────────
        outBuf.rewind()
        interp.runForMultipleInputsOutputs(
            arrayOf(inBuf),
            hashMapOf(detectionOutputIdx to outBuf) as Map<Int, Any>
        )
        outBuf.rewind()

        // ── 解析輸出 [5, 8400] ──────────────────────────────────
        fun getValue(ch: Int, anchor: Int): Float =
            if (outIsFloat) outBuf.getFloat((ch * 8400 + anchor) * 4)
            else (outBuf.get(ch * 8400 + anchor).toInt() - outZeroPoint) * outScale

        // ── 偵測模型輸出座標格式（normalized [0,1] 或 pixel [0,640]）───
        // 模型格式不會改變，只在第一次確認，之後直接用快取值（省 8400 次 ByteBuffer 讀取/幀）
        val coordScale: Float
        if (cachedCoordScale != null) {
            coordScale = cachedCoordScale!!
        } else {
            var maxConfForScale = 0f; var maxCxForScale = 0f
            for (i in 0 until 8400) {
                val c = getValue(4, i)
                if (c > maxConfForScale) { maxConfForScale = c; maxCxForScale = getValue(0, i) }
            }
            coordScale = if (maxConfForScale > 0.05f && maxCxForScale < 2.0f) INPUT_SIZE.toFloat() else 1.0f
            if (maxConfForScale > 0.05f) {  // 有足夠信心才鎖定格式
                cachedCoordScale = coordScale
                Log.d(TAG, "coordScale locked=$coordScale (maxConf=${"%.4f".format(maxConfForScale)} maxCx_raw=${"%.4f".format(maxCxForScale)})")
            }
        }

        // debug：每 20 次 detect 完整掃描（僅 debug build，避免 release 浪費 8400 次讀取）
        detectCallCount++
        if (BuildConfig.DEBUG && detectCallCount % 20 == 0) {
            val coordFormat = if (coordScale > 1f) "normalized→scaled" else "pixel"
            val LOW_THRESH = 0.10f
            data class Det(val conf: Float, val cx: Float, val cy: Float, val w: Float, val h: Float)
            val topDets = mutableListOf<Det>()
            for (i in 0 until 8400) {
                val c = getValue(4, i)
                val cx = getValue(0, i) * coordScale
                val cy = getValue(1, i) * coordScale
                if (c >= LOW_THRESH) topDets.add(Det(c, cx, cy, getValue(2,i) * coordScale, getValue(3,i) * coordScale))
            }
            topDets.sortByDescending { it.conf }
            val topStr = topDets.take(5).joinToString(" | ") {
                val inEdge = it.cx < tileEdgeMargin || it.cy < tileEdgeMargin ||
                             it.cx > INPUT_SIZE - tileEdgeMargin || it.cy > INPUT_SIZE - tileEdgeMargin
                "conf=${"%.3f".format(it.conf)} tile=(${"%.1f".format(it.cx)},${"%.1f".format(it.cy)}) ${if (inEdge) "EDGE" else "OK"}"
            }
            Log.d(TAG, "[debug#$detectCallCount] roi=(${roiCenterX},${roiCenterY}) tile=($tileLeft,$tileTop) coordFormat=$coordFormat")
            if (topDets.isNotEmpty()) {
                Log.d(TAG, "[debug#$detectCallCount] top-${minOf(5, topDets.size)} above 0.10: $topStr")
            } else {
                Log.d(TAG, "[debug#$detectCallCount] ⚠️ 無任何偵測超過 0.10")
            }
        }

        // ── 篩選 + 座標轉換回 frame 空間 ────────────────────────
        val scaleX = tileW.toFloat() / INPUT_SIZE
        val scaleY = tileH.toFloat() / INPUT_SIZE
        val dets = mutableListOf<FloatArray>()
        for (i in 0 until 8400) {
            val conf = getValue(4, i)
            if (conf < confThreshold) continue

            // 模型輸出座標：依 coordScale 統一轉為 pixel [0, 640]
            val cx_tile = getValue(0, i) * coordScale
            val cy_tile = getValue(1, i) * coordScale
            val w_tile  = getValue(2, i) * coordScale
            val h_tile  = getValue(3, i) * coordScale

            // 排除 tile 邊緣的 false positive（YOLOv8 anchor grid 在邊界常產生噪訊）
            // tileEdgeMargin 由呼叫端控制：正常=24f，擊球後=8f（允許高速球靠邊緣）
            if (cx_tile < tileEdgeMargin || cy_tile < tileEdgeMargin ||
                cx_tile > INPUT_SIZE - tileEdgeMargin ||
                cy_tile > INPUT_SIZE - tileEdgeMargin) continue

            // 轉回 frame 座標
            val cx = tileLeft + cx_tile * scaleX
            val cy = tileTop  + cy_tile * scaleY
            val w  = w_tile   * scaleX
            val h  = h_tile   * scaleY
            dets.add(floatArrayOf(cx, cy, w, h, conf))
        }

        return nonMaxSuppression(dets)
    }

    // ──────────────────────────────────────────────────────────────

    private fun findDetectionOutputTensor(): Pair<Int, org.tensorflow.lite.Tensor>? {
        val interp = interpreter ?: return null
        val n = interp.outputTensorCount
        for (i in 0 until n) {
            val t = interp.getOutputTensor(i)
            val s = t.shape()
            if (s.size == 3 && s[1] == 5 && s[2] == 8400) return Pair(i, t)
        }
        for (i in 0 until n) {
            val t = interp.getOutputTensor(i)
            val s = t.shape()
            if (s.size == 3 && s[2] == 8400) return Pair(i, t)
        }
        return null
    }

    private fun nonMaxSuppression(dets: List<FloatArray>): List<FloatArray> {
        if (dets.isEmpty()) return emptyList()
        val sorted = dets.sortedByDescending { it[4] }
        val suppressed = BooleanArray(sorted.size)
        val kept = mutableListOf<FloatArray>()
        for (i in sorted.indices) {
            if (suppressed[i]) continue
            kept.add(sorted[i])
            for (j in i + 1 until sorted.size) {
                if (!suppressed[j] && iou(sorted[i], sorted[j]) > NMS_IOU) suppressed[j] = true
            }
        }
        return kept
    }

    private fun iou(a: FloatArray, b: FloatArray): Float {
        val ax1 = a[0] - a[2] / 2f; val ay1 = a[1] - a[3] / 2f
        val ax2 = a[0] + a[2] / 2f; val ay2 = a[1] + a[3] / 2f
        val bx1 = b[0] - b[2] / 2f; val by1 = b[1] - b[3] / 2f
        val bx2 = b[0] + b[2] / 2f; val by2 = b[1] + b[3] / 2f
        val ix1 = max(ax1, bx1); val iy1 = max(ay1, by1)
        val ix2 = min(ax2, bx2); val iy2 = min(ay2, by2)
        if (ix2 <= ix1 || iy2 <= iy1) return 0f
        val inter = (ix2 - ix1) * (iy2 - iy1)
        return inter / ((ax2-ax1)*(ay2-ay1) + (bx2-bx1)*(by2-by1) - inter)
    }

    fun close() {
        interpreter?.close(); interpreter = null
        inputBuf = null; outputBuf = null
    }
}
